src/analyses/constant_propagator.cpp

```diff
@@ -761,37 +761,43 @@ void constant_propagator_ait::replace(
     {
       exprt c = it->get_condition();
       replace_types_rec(d.values.replace_const, c);
-      constant_propagator_domaint::partial_evaluate(d.values, c, ns);
-      it->set_condition(c);
+      if(!constant_propagator_domaint::partial_evaluate(d.values, c, ns))
+        it->set_condition(c);
     }
     else if(it->is_assign())
     {
-      exprt &rhs=to_code_assign(it->code).rhs();
-      constant_propagator_domaint::partial_evaluate(d.values, rhs, ns);
+      auto assign = it->get_assign();
+      exprt &rhs = assign.rhs();
 
-      if(rhs.id()==ID_constant)
-        rhs.add_source_location() =
-          to_code_assign(it->code).lhs().source_location();
+      if(!constant_propagator_domaint::partial_evaluate(d.values, rhs, ns))
+      {
+        if(rhs.id() == ID_constant)
+          rhs.add_source_location() = assign.lhs().source_location();
+        it->set_assign(assign);
+      }
     }
     else if(it->is_function_call())
     {
-      exprt &function = to_code_function_call(it->code).function();
-      constant_propagator_domaint::partial_evaluate(d.values, function, ns);
+      auto call = it->get_function_call();
 
-      exprt::operandst &args=
-        to_code_function_call(it->code).arguments();
+      constant_propagator_domaint::partial_evaluate(
+        d.values, call.function(), ns);
 
-      for(auto &arg : args)
-      {
+      for(auto &arg : call.arguments())
         constant_propagator_domaint::partial_evaluate(d.values, arg, ns);
-      }
+
+      it->set_function_call(call);
```

Comment:
This one could also be made conditional, though it requires an additional Boolean to track whether any modification took place.

Reply:
I did contemplate it, and arrived at the conclusion that the "set" is likely cheaper than the Boolean logic.

Reply:
The `set_function_call` might be cheap, but the problem with breaking sharing is that you pay the price multiple times: every time the expression is stored in a `std::unordered_set` or `std::unordered_map` the `operator==` becomes more expensive, because pointer-comparison cannot be used. There is of course a question whether this expression is shared, but it quite possibly is if the function has multiple call sites. Destroying sharing will thus make writing the simplified goto binary more expensive.

Reply:
ok, now with boolean logic